### PR TITLE
Faster CLR Calculation

### PR DESCRIPTION
**Pull request by frankchen07**
### Description

Added three methods:

1. method that translates grant dictionary data into a list of lists
2. method that aggregates and calculates the by pair totals
3. method that continually tests thresholds until the proper total CLR amount is found, and calculates the result

### Testing

I was going through the curve calculation part, and I believe it may break it. My concern is around line 224, but it may be that it's alright. Let's pair up and take look at that @thelostone-mc.

### CODE REVIEW

**Comment by owocki** [comment]
how much faster is this?

File: app/grants/clr.py [hunk 1]
```diff
@@ -47,125 +47,113 @@
...
-        'id': grant_id,
-        'profile_pairs': profile_pairs,
-        'contribution_pairs': contribution_pairs,
-        'sqrt_of_product_pairs': sqrt_of_product_pairs
-    }
-
-    return grant
+def translate_data(grants_data):
+    grants_list = []
+    for g in grants_data:
```
Comment (thelostone-mc) [inline app/grants/clr.py]: `g -> grant`

File: app/grants/clr.py [hunk 2]
```diff
@@ -47,125 +47,113 @@
...
-
-    grant = {
-        'id': grant_id,
-        'profile_pairs': profile_pairs,
-        'contribution_pairs': contribution_pairs,
-        'sqrt_of_product_pairs': sqrt_of_product_pairs
-    }
-
-    return grant
+def translate_data(grants_data):
```
Comment (thelostone-mc) [inline app/grants/clr.py]: `grants_data -> grants`

File: app/grants/clr.py [hunk 3]
```diff
@@ -47,125 +47,113 @@
...
+    grants_list = []
+    for g in grants_data:
+        grant_id = g.get('id')
+        for c in g.get('contributions'):
```
Comment (thelostone-mc) [inline app/grants/clr.py]: `c -> contribution`

File: app/grants/clr.py [hunk 4]
```diff
@@ -47,125 +47,113 @@
...
+def aggregate_contributions(grant_contributions):
+    contrib_dict = {}
+    for proj, user, amount in grant_contributions:
+        if proj not in contrib_dict:
+            contrib_dict[proj] = {}
+        contrib_dict[proj][user] = contrib_dict[proj].get(user, 0) + amount
+
+    tot_overlap = {}
+    for proj, contribz in contrib_dict.items():
+        for k1, v1 in contribz.items():
```
Comment (thelostone-mc) [inline app/grants/clr.py]: should we rename k1 , v1 into something more meaningful

File: app/grants/clr.py [hunk 5]
```diff
@@ -47,125 +47,113 @@
...
+            if k1 not in tot_overlap:
+                tot_overlap[k1] = {}
+            for k2, v2 in contribz.items():
```
Comment (thelostone-mc) [inline app/grants/clr.py]: should we rename k2 , v2 into something more meaningful

File: app/grants/clr.py [hunk 6]
```diff
@@ -47,125 +47,113 @@
...
+            tot = 0
+            for k1, v1 in contribz.items():
+                for k2, v2 in contribz.items():
+                    if k2 > k1:  # remove pairs
+                        # # pairwise matching formula
+                        # tot += (v1 * v2) ** 0.5 * min(1, threshold / pair_totals[k1][k2])
+                        # vitalik's division formula
+                        tot += ((v1 * v2) ** 0.5) / (pair_totals[k1][k2] / threshold + 1)
+            bigtot += tot
+            # totals.append((proj, tot))
```
Comment (thelostone-mc) [inline app/grants/clr.py]: if it's not needed -> let's remove it

**Comment by thelostone-mc** [comment]
@owocki so 
@frankchen07 did a run here https://github.com/thelostone-mc/clr/tree/frank-dev

Ran both the algo against the previous clr dataset
algorithm: ran 1 time
data: previous clr data set from live
speed: ~5x faster

Upside: 
- this is 5x faster

Reality: 
- it's not still fast enough to be able to run it on demand and has to run has a cron
- it can be run ~30times as opposed to ~6 times per day which makes the results a lil more accurate but not something the user will notice
- the end results of both algo are the same 


cc @gitcoinco/engineers

**Pull request merged**